### ISSUE TITLE
Enhancement: Build against .NET 4.6 and .NET Core

### ISSUE THREAD

**Issue by CVertex** [JohnnyCrazy/SpotifyAPI-NET#64]
Microsoft is ending support for .net framework 4.5 in January.. I have no idea why.. I think they're trying to push all 4.5 apps to 4.6..

Since the only real hard dependency is json.net, maybe we could upgrade this library to support both .NET 4.6 and .NET Core so we can run this cross platform.

I'm looking forward to running this library on a linux server :smile: 

Thanks

**Comment by JohnnyCrazy** [comment]
Generally, I like the idea of porting the project to 4.6.
But I don't want to step into JSON.NET too much. 

>maybe we could upgrade this library

What is your plan on upgrading it? Creating a PR on JSON.NET's repo or creating an own version? I would definitely go for the first approach since I would like to stick to the main JSON.NET project.

**Comment by CVertex** [comment]
Apologies, by "this library", I meant SpotifyAPI.NET. Not Json.NET

Json.NET supports .NET Core, the upgrade should be smooth as SpotifyAPI has few dependencies. 

WebClient is also part of .NET Core.

**Comment by JohnnyCrazy** [comment]
Alright.

I ran some tests, and it seems like ít needs some (more) work.
You can find the detailed report here: [dl.johnnycrazy.de/spotifyapi/ApiPortAnalysis.html](http://dl.johnnycrazy.de/spotifyapi/ApiPortAnalysis.html)

While I think it's realizable, I would consider it as a major change and propably needs some time.

**Comment by CVertex** [comment]
@JohnnyCrazy Nice. Looks like you ran the whole report on the whole solution! You only need to run it on the SpotifyAPI.NET library project. WinForms + WPF aren't coming to .NET Core for a long time, so it's reasonable to have the following goals:

* SpotifyAPI-Net Library - deps on 4.6.1 and  .NET Core 5
* SpotifyAPI-Net winforms app - just depends on 4.6.1

**Comment by CVertex** [comment]
@JohnnyCrazy My bad, I forgot about the local API. There's lots of windows specific stuff in there. Yeah, this will require some effort

**Comment by Mrnikbobjeff** [comment]
I actually attended the Microsoft Developer conference where they introduced .Net Core and AFAIK it mainly just boosts the performance. Concerning the end of the .NET 4.5.2 support, most previous versions are absolutely fine, just new updates will not reach these versions. Also I would not advise to switch to .NET 4.6 so easily as it still is a very new technology. The people at StackOverflow wanted to switch earlier when .NET 4.6 just came out and they were really suffering. Bugs such as Tailspin or similar were only found by programmers like who were furiously working to find the error in their code. Because of things like this and considering that this is a very lightweight library I do not think that we have to switch versions.

**Comment by CVertex** [comment]
Thanks for your comments @Mrnikbobjeff 

> Also I would not advise to switch to .NET 4.6 so easily as it still is a very new technology. The people at StackOverflow wanted to switch earlier when .NET 4.6 just came out and they were really suffering. Bugs such as Tailspin or similar were only found by programmers like who were furiously working to find the error in their code.

Interesting..
http://nickcraver.com/blog/2015/07/27/why-you-should-wait-on-dotnet-46/

Makes sense to avoid this for production for most applications. However, for our apps, we're okay with breaking some things to use the new tech. I'm very keen to migrate my codebase to .NET core for example.

Perhaps we can create/maintain a fork that builds against .NET framework 4.6. The upgrade should be a simple switch in csproj.

Regarding .NET core, I spoke to a MS colleague here. He suggests disabling the Spotify Local API when building for .NET core. Of course, if you're app is a full framework app, you'll get access to local API.

I'm very interested in having this library on .NET core in the coming months so I can run my crawlers on linux.

**Comment by erinxocon** [comment]
I'm currently attempting to port this to .NET core and I've managed to have some success with the web api stuff. 

I'm running into two dependency issues:

```
/Users/erin/repos/SpotifyAPI-NET/SpotifyAPI/src/Web/SimpleHttpServer.cs(12,14): error CS0234: The type or namespace name 'Web' does not exist in the namespace 'System' (are you missing an assembly reference?)                                                                              
/Users/erin/repos/SpotifyAPI-NET/SpotifyAPI/src/Web/SpotifyWebClient.cs(17,26): error CS0246: The type or namespace name 'WebClient' could not be found (are you missing a using directive or an assembly reference?) 
```

I thought WebClient was supported by .NET core but I can't seem to get it to work.  Any ideas?  If I have to I can try and rework these things into something that's supported in .NET core.  I'm relatively new to .NET core, thought porting this library would be an interesting challenge!

**Comment by jholzer** [comment]
.Net Core is not a 1:1 implementation of the full .Net-framework.
Have a look at this:

http://stackoverflow.com/questions/30227219/how-to-use-webclient-with-netcore

**Comment by erinxocon** [comment]
Yeah I know. When I first started porting the Web library I had 11 dependency issues. Resolved all of those except those two. I'll look into the WebClient stuff tonight from that link you sent. Thanks!

**Comment by erinxocon** [comment]
@JohnnyCrazy I can use HttpClient in `SpotifyWebClient.cs`, but just to clarify, `SimpleHttpServer.cs` is just a little http server used for internal stuff?  Still trying to wrap my mind around how the whole thing works.

**Comment by erinxocon** [comment]
Annnndddd after some investigation HttpClient by itself doesn't offer enough of the functionality WebClient did, and I'm not knowledgable enough in C# libraries to accomplish it!  Perhaps someone else can help, they can fork my [branch](https://github.com/erinxocon/SpotifyAPI-NET/tree/dotnetcore) and submit a pull request!

**Comment by erinxocon** [comment]
So I know we closed the splitting up of the projects, but I'm still curious as to where we were on a rewrite of the Http Server component.  What kinds of auth do we need it for?  I might be able to work on some of those?

@JohnnyCrazy @petrroll

**Comment by JohnnyCrazy** [comment]
@erinxocon 
I will open an issue with my thoughts on how it should work tomorrow. Then you are free to go for it if you want to! :)

**Comment by erinxocon** [comment]
@JohnnyCrazy Thanks!  Looking for something to hack on as I'm in between projects right now.

**Comment by petrroll** [comment]
@JohnnyCrazy Looking forward to it as well :).

**Comment by JohnnyCrazy** [comment]
Hi, 

.NET Standard support has landed in 3.0.0

Please use the following NuGet Packages:
[SpotifyAPI.Web](https://www.nuget.org/packages/SpotifyAPI.Web/) - The API Client
[SpotifyAPI.Web.Auth](https://www.nuget.org/packages/SpotifyAPI.Web.Auth/) - The Auth flows

Docs will be updated soon!
